### ISSUE TITLE
New PHP code generator

### ISSUE THREAD

**Issue by murtukov** [overblog/GraphQLBundle#668]
As we know the GraphQL Bundle relies in it's core on an internal PHP code generator library. When I was implementing the validation feature I ran into a problem of editing the generator code. The library is not meant to be scaled and editing it was a hard task. Currently the code under `src/Generator` and `lib/generator` is pretty messy and it's hard to follow, how it works. I am planning to work on other features, which require even more changes of the generator code, that's why I decided to write a new PHP code generator library first. It should clean the current mess, make introduction of new features (depending on the generator) much easier and, if possible, optimize the performance. 

The library is already finished and now I am on the step of integrating it into the bundle. But before I do it, there are some questions, that need to be adressed first. The generator is the core of the bundle and changing it requires attention to other aspects, for example config preprocessing and shape of the config tree, which the new php generator will rely on.

First of all I would like to draw your attention to the config tree of yaml types. Here are some random type configs:

```yaml
DateTime:
    type: custom-scalar
    config:
        description: "The DateTime type"
        serialize: ["Overblog\\GraphQLGenerator\\Tests\\DateTimeType", "serialize"]
        parseValue: ["Overblog\\GraphQLGenerator\\Tests\\DateTimeType", "parseValue"]
        parseLiteral: ["Overblog\\GraphQLGenerator\\Tests\\DateTimeType", "parseLiteral"]

Character:
    type: interface
    config:
        description: "@='A character' ~ ' in the Star Wars Trilogy'"
        fields:
            id:
                type: "String!"
                description: "The id of the character."
            name:
                type: "String"
                description: "The name of the character."
            friends:
                type: "[Character]"
                description: "The friends of the character."
            appearsIn:
                type: "[Episode]"
                description: "Which movies they appear in."
        # used expression language to defined resolver (tagged services)
        resolveType: "Overblog\\GraphQLGenerator\\Tests\\Resolver::resolveType"

Droid:
    type: object
    config:
        description: "A mechanical creature in the Star Wars universe."
        fields:
            id:
                type: "String!"
                description: "The id of the droid."
                resolve: "@=value['id']"
            name:
                type: "String"
                description: "The name of the droid."
            friends:
                type: "[Character]"
                description: "The friends of the droid, or an empty list if they have none."
                resolve: ["Overblog\\GraphQLGenerator\\Tests\\Resolver" , "getFriends"]
            appearsIn:
                type: "[Episode]"
                description: "Which movies they appear in."
            primaryFunction:
                type: "String"
                description: "The primary function of the droid."
        interfaces: [Character]
```
As you can see there is always the entry `config` which I personally find absolutely unnecessary, it just adds a gap to each type. I know, that the original intention was to keep everything related to `webonyx/graphql-php` under the `config` and everything else 1 level higher, but it's not relevant anymore, because:

1) We have validation, which introduces the `validation` key, which doesn't belong to `webonyx/graphql-php` configuration
2) From the point of view of users the types are configs themselves, and adding the entry `config` doesn't make much sense.

So my first question is: **Should we remove the `config` key?** My vote is **yes**. It needs to be decided before the version 1.0, as it is a breaking change. I'm going to wait, what others say. 

Here are the same configs, but without the `config` key
```yaml
DateTime:
    type: custom-scalar
    description: "The DateTime type"
    serialize: ["Overblog\\GraphQLGenerator\\Tests\\DateTimeType", "serialize"]
    parseValue: ["Overblog\\GraphQLGenerator\\Tests\\DateTimeType", "parseValue"]
    parseLiteral: ["Overblog\\GraphQLGenerator\\Tests\\DateTimeType", "parseLiteral"]

Character:
    type: interface
    description: "@='A character' ~ ' in the Star Wars Trilogy'"
    fields:
        id:
            type: "String!"
            description: "The id of the character."
        name:
            type: "String"
            description: "The name of the character."
        friends:
            type: "[Character]"
            description: "The friends of the character."
        appearsIn:
            type: "[Episode]"
            description: "Which movies they appear in."
    # used expression language to defined resolver (tagged services)
    resolveType: "Overblog\\GraphQLGenerator\\Tests\\Resolver::resolveType"

Droid:
    type: object
    description: "A mechanical creature in the Star Wars universe."
    fields:
        id:
            type: "String!"
            description: "The id of the droid."
            resolve: "@=value['id']"
        name:
            type: "String"
            description: "The name of the droid."
        friends:
            type: "[Character]"
            description: "The friends of the droid, or an empty list if they have none."
            resolve: ["Overblog\\GraphQLGenerator\\Tests\\Resolver" , "getFriends"]
        appearsIn:
            type: "[Episode]"
            description: "Which movies they appear in."
        primaryFunction:
            type: "String"
            description: "The primary function of the droid."
    interfaces: [Character]
```
---

The second question is **Why do we use the identifier `input-object` instead of just `input`**? If there is no particular reason for that, I suggest to deprecate it and introduce `input`

---
The new generator library will also reduce the size of generated classes, thanks to the arrow functions, introduced in PHP 7.4 and by removing of null/empty entries, e.g.:

Before:
```php
final class RootQueryType extends ObjectType implements GeneratedTypeInterface
{
    const NAME = 'RootQuery';

    public function __construct(ConfigProcessor $configProcessor, GlobalVariables $globalVariables = null)
    {
        $configLoader = function(GlobalVariables $globalVariable) {
            return [
                'name' => 'RootQuery',
                'description' => null,
                'fields' => function () use ($globalVariable) {
                    return [
                        'object' => [
                            'type' => $globalVariable->get('typeResolver')->resolve('ObjectWithPrivateField'),
                            'args' => [],
                            'resolve' => function ($value, $args, $context, ResolveInfo $info) use ($globalVariable) {
                                return ["name" => "His name"];
                            },
                            'description' => null,
                            'deprecationReason' => null,
                            'complexity' => null,
                            # public and access are custom options managed only by the bundle
                            'public' => null,
                            'access' => null,
                            'useStrictAccess' => true,
                        ],
                    ];
                },
                'interfaces' => function () use ($globalVariable) {
                    return [];
                },
                'isTypeOf' => null,
                'resolveField' => null,
            ];
        };
        $config = $configProcessor->process(LazyConfig::create($configLoader, $globalVariables))->load();
        parent::__construct($config);
    }
}
```

After:
```php
final class RootQueryType extends ObjectType implements GeneratedTypeInterface
{
    const NAME = 'RootQuery';

    public function __construct(ConfigProcessor $configProcessor, GlobalVariables $globalVariables = null)
    {
        $configLoader = fn() => [
            'name' => self::NAME,
            'fields' => fn() => [
                'object' => [
                    'type' => $globalVariables->get('typeResolver')->resolve('ObjectWithPrivateField'),
                    'resolve' => fn() => ["name" => "His name"],
                ],
            ],
        ];
        $config = $configProcessor->process(LazyConfig::create($configLoader, $globalVariables))->load();
        parent::__construct($config);
    }
}
```

The content of the `resolve` will be generated accordingly, it's not possible to always use arrow function. Maybe it would be also good, to replace the GLobalVariable with separate objects to avoid chain calls like that:
```php
$globalVariables->get('typeResolver')->resolve('ObjectWithPrivateField')
```
and use instead:
```php
$typeResolver->get('ObjectWithPrivateField')
```

**Comment by akomm** [comment]
1. config key: agree. but can not tell if that has some negative implications to any feature.
2. input-object: agree
3. nulls removed: while undefined index access resolved to null, the behavior is different, when you have null set explicitly. Did you test if this change would break something? Otherwise agree.

**Comment by murtukov** [comment]
@akomm 

> Did you test if this change would break something?

Sure. It work's perfectly without null entries and it's totally expected. If you use `webonyx/graphql-php` directly you are supposed to create this configs manually and it would be strange if you were required to explicitly enter options you don't need.
